# Method chaining

# Why method chaining?

Chaining lets you instantiate, configure, run, edit or apply any method all in one readable pipeline.
In apsimNGpy, most builder methods return self (so you can keep chaining), while terminal methods like ``get_simulated_output`` return results (e.g., a DataFrame) and end the chain. See the example below.

In [7]:
from apsimNGpy.core.apsim import ApsimModel
df =ApsimModel('Maize').run().get_simulated_output('Report')
df.mean(numeric_only=True)

CheckpointID                     1.000000
SimulationID                     1.000000
Maize.AboveGround.Wt          1225.099950
Maize.AboveGround.N             12.381196
Yield                         5636.529504
Maize.Grain.Wt                 563.652950
Maize.Grain.Size                 0.284941
Maize.Grain.NumberFunction    1986.770519
Maize.Grain.Total.Wt           563.652950
Maize.Grain.N                    7.459296
Maize.Total.Wt                1340.837427
dtype: float64

# Method chaining with some model editing

In [5]:
df = ApsimModel('Maize').edit_model(model_type='Models.Manager', 
                                    model_name='Sow using a variable rule', 
                                    Population =12).run().results
df.mean(numeric_only=True)

CheckpointID                     1.000000
SimulationID                     1.000000
Maize.AboveGround.Wt          1489.537802
Maize.AboveGround.N             14.744787
Yield                         6401.875060
Maize.Grain.Wt                 640.187506
Maize.Grain.Size                 0.285645
Maize.Grain.NumberFunction    2327.909542
Maize.Grain.Total.Wt           640.187506
Maize.Grain.N                    8.455346
Maize.Total.Wt                1620.941414
dtype: float64

Just to make sure it is working, let's change the population to 4

In [6]:
df = ApsimModel('Maize').edit_model(model_type='Models.Manager',
                                    model_name='Sow using a variable rule',
                                    Population =4).run().results
df.mean(numeric_only=True)

CheckpointID                     1.000000
SimulationID                     1.000000
Maize.AboveGround.Wt          1039.267466
Maize.AboveGround.N             10.433845
Yield                         4786.612059
Maize.Grain.Wt                 478.661206
Maize.Grain.Size                 0.284626
Maize.Grain.NumberFunction    1661.918597
Maize.Grain.Total.Wt           478.661206
Maize.Grain.N                    6.298505
Maize.Total.Wt                1136.537577
dtype: float64

Since example one and two produced different mean values, it implies that our editing worked. For example, mean maize yield at a population of `12`, was
6401.875060 kg while at 4 was 4786.612059 kg. Method chaining, however, has some drawbacks such as poor readability, and maybe harder to debug chained methods